pylint/config/config_initialization.py

```diff
@@ -141,7 +145,8 @@ def _config_initialization(
     linter._parse_error_mode()
 
     # Link the base Namespace object on the current directory
-    linter._directory_namespaces[Path(".").resolve()] = (linter.config, {})
+    if len(linter._directory_namespaces) == 0:
+        linter._directory_namespaces[Path(".").resolve()] = (linter.config, {})
```

Comment:
For now I'd prefer to revert the changes in these two lines.

They are really rightly coupled to the final implementation of the per directory configs and the performance impact is hard to judge on its own. As far as I can see, all other changes in this PR are somewhat sensible on their own. This one isn't.